tools/mkcomposefs.c

```diff
@@ -893,43 +877,318 @@ static ssize_t write_cb(void *_file, void *buf, size_t count)
+		return 0;
+
+	size_t n_children = lcfs_node_get_n_children(node);
+	for (size_t i = 0; i < n_children; i++) {
+		struct lcfs_node_s *child = lcfs_node_get_child(node, i);
+		if (construct_compute_data(child, collection, path) < 0) {
+			return -1;
+		}
 	}
 
 	return 0;
 }
 
+struct work_item_iterator {
+	pthread_mutex_t mutex_node_iterator;
+	int current_item;
+	int errorcode;
+	bool cancel_request;
+};
+
+static struct work_item *get_next_work_item(struct work_collection *collection,
+					    struct work_item_iterator *iterator)
+{
+	if (!iterator || !collection)
+		return NULL;
+
+	bool cancel = false;
+	struct work_item *ret = NULL;
+
+	pthread_mutex_lock(&(iterator->mutex_node_iterator));
+	if (iterator->cancel_request)
+		cancel = true;
+	else if (iterator->current_item < collection->count) {
+		ret = &(collection->items[iterator->current_item]);
+		iterator->current_item++;
+	}
+	pthread_mutex_unlock(&(iterator->mutex_node_iterator));
+	return cancel ? NULL : ret;
+}
+
+static void request_cancel(struct work_item_iterator *iterator, int errorcode)
+{
+	pthread_mutex_lock(&(iterator->mutex_node_iterator));
+	iterator->cancel_request = true;
+	iterator->errorcode = errorcode;
+	pthread_mutex_unlock(&(iterator->mutex_node_iterator));
+}
+
+typedef int (*THREAD_PROCESS_PROC)(struct work_item *, void *);
+
+static int process_copy(struct work_item *item, void *digest_store_path)
+{
+	if (copy_file_with_dirs_if_needed(item->path, (const char *)digest_store_path,
+					  lcfs_node_get_payload(item->node),
+					  true) < 0) {
+		return -1;
+	}
+	return 0;
+}
+
+static int process_compute(struct work_item *item, void *data)
+{
+	int buildflag = (int)(long)data;
+	return lcfs_node_set_from_content(item->node, AT_FDCWD, item->path, buildflag);
+}
+
+struct thread_data {
+	THREAD_PROCESS_PROC proc;
+	struct work_collection *collection;
+	struct work_item_iterator *iterator;
+	void *data;
+};
+
+static void *thread_proc(void *data)
+{
+	struct thread_data *info = (struct thread_data *)data;
+
+	while (true) {
+		struct work_item *item =
+			get_next_work_item(info->collection, info->iterator);
+
+		if (!item)
+			return 0;
+
+		if (!item->node) {
+			request_cancel(info->iterator, EINVAL);
+			return 0;
+		}
+
+		if (info->proc(item, info->data) != 0) {
+			request_cancel(info->iterator, errno);
+			return 0;
+		}
+	}
+	return 0;
+}
+
+static int execute_in_threads(const int requested_threads,
+			      struct work_collection *collection,
+			      THREAD_PROCESS_PROC proc, void *data)
+{
+	struct work_item_iterator iterator;
+	int ret = pthread_mutex_init(&iterator.mutex_node_iterator, NULL);
+	if (0 != ret) {
+		errno = ret;
+		return -1;
+	}
+
+	iterator.current_item = 0;
+	iterator.errorcode = 0;
+	iterator.cancel_request = false;
+
+	struct thread_data thread_info;
+	thread_info.data = data;
+	thread_info.proc = proc;
+	thread_info.collection = collection;
+	thread_info.iterator = &iterator;
+
+	pthread_t *threads = NULL;
```

Comment:
I would use:
cleanup_free phread_t *threads = NULL;

then you can drop all calls to free(threads)

Reply:
fixed